crates/iceberg-catalog/src/catalog/io.rs

```diff
@@ -1,14 +1,110 @@
 use iceberg_ext::catalog::rest::IcebergErrorResponse;
 use serde::Serialize;
+use std::collections::HashMap;
 use std::io::Write;
 
+use super::CommonMetadata;
+
+#[derive(thiserror::Error, Debug, strum::IntoStaticStr)]
+pub enum CompressionCodecError {
+    #[error("Unsupported compression codec: {0}")]
+    UnsupportedCompressionCodec(String),
+}
+
+impl CompressionCodecError {
+    pub fn to_type(&self) -> &'static str {
+        self.into()
+    }
+}
+
+impl From<CompressionCodecError> for IcebergErrorResponse {
+    fn from(value: CompressionCodecError) -> Self {
+        let typ = value.to_type();
+        let boxed = Box::new(value);
+        let message = boxed.to_string();
+
+        match value {
+            CompressionCodecError::UnsupportedCompressionCodec(_) => {
+                ErrorModel::bad_request(message, typ, Some(boxed)).into()
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum CompressionCodec {
+    None,
+    Gzip,
+}
+
+impl CompressionCodec {
+    pub fn compress(self, payload: &[u8]) -> Result<Vec<u8>, IoError> {
+        match self {
+            CompressionCodec::None => Ok(payload.to_vec()),
+            CompressionCodec::Gzip => {
+                let mut compressed_metadata = GzEncoder::new(Vec::new(), Compression::default());
+                compressed_metadata
+                    .write_all(payload)
+                    .map_err(IoError::FileCompression)?;
+
+                compressed_metadata
+                    .finish()
+                    .map_err(IoError::FileCompression)
+            }
+        }
+    }
+
+    pub fn as_file_extension(self) -> &'static str {
+        match self {
+            CompressionCodec::None => "",
+            CompressionCodec::Gzip => ".gz",
+        }
+    }
+
+    pub fn try_from_properties(
+        properties: &HashMap<String, String>,
+    ) -> Result<Self, CompressionCodecError> {
+        properties
+            .get(METADATA_COMPRESSION)
+            .map(String::as_str)
+            .map_or(Ok(Self::default()), |value| match value {
+                "gzip" => Ok(Self::Gzip),
+                "none" => Ok(Self::None),
+                unknown => Err(CompressionCodecError::UnsupportedCompressionCodec(
+                    unknown.into(),
+                )),
+            })
+    }
+
+    pub fn try_from_maybe_properties(
+        maybe_properties: Option<&HashMap<String, String>>,
+    ) -> Result<Self, CompressionCodecError> {
+        match maybe_properties {
+            Some(properties) => Self::try_from_properties(properties),
+            None => Ok(Self::default()),
+        }
+    }
+
+    pub fn try_from_metadata<T: CommonMetadata>(
+        metadata: &T,
+    ) -> Result<Self, CompressionCodecError> {
+        Self::try_from_properties(metadata.properties())
+    }
+}
+
+impl Default for CompressionCodec {
+    fn default() -> Self {
+        Self::Gzip
+    }
+}
```

Comment:
Thank you for that hint, I wasn't aware of that possibility. I did not get that suggestion from clippy and my IDE setup (nvim + rust-analyzer + rustaceanvim) should suggest something like this. Weird, because it absolutely makes sense to use that feature.